File: scripts/auto_bump.py
import argparse
import os
import re
import subprocess
from typing import List

def get_workspace_members() -> List[str]:
    """Reads Cargo.toml to find workspace members."""
    cargo_toml_path = os.path.join(os.getcwd(), 'Cargo.toml')
    if not os.path.exists(cargo_toml_path):
        raise FileNotFoundError(f"Cargo.toml not found at {cargo_toml_path}")

    with open(cargo_toml_path, 'r') as f:
        content = f.read()

    members_match = re.search(r'\[workspace\]\nmembers = \[(.*?)\]', content, re.DOTALL)
    if not members_match:
        # Fallback for single-package Cargo.toml or different format
        return []
    
    # Extract members, clean up whitespace and quotes
    members = [m.strip().strip('"') for m in members_match.group(1).split(',') if m.strip()]
    
    # Filter for actual directories relative to the workspace root
    return [m for m in members if os.path.isdir(m)]

def find_version_files(workspace_members: List[str]) -> List[str]:
    """Finds all Cargo.toml files in the workspace and relevant UI/doc files."""
    files = []
    # Cargo.toml files
    files.append('Cargo.toml') # Root Cargo.toml
    for member in workspace_members:
        files.append(os.path.join(member, 'Cargo.toml'))
    
    # UI/Doc files (hardcoded paths as they are specific to this project's structure)
    files.append('docs/MANUAL.md')
    files.append('installer-cli/src/tui/render.rs')
    files.append('installer-cli/src/tui/menus.rs')
    files.append('docs/HISTORY.md')
    files.append('rust-toolchain.toml') # Added for checking purposes, though version is not directly in it

    # Filter out files that don't exist to prevent errors
    return [f for f in files if os.path.exists(f)]

def get_current_version(file_path: str) -> str:
    """Extracts the version from a Cargo.toml or rust-toolchain.toml file."""
    with open(file_path, 'r') as f:
        content = f.read()
        match_cargo = re.search(r'version\s*=\s*"(\d+\.\d+\.\d+)"', content)
        if match_cargo:
            return match_cargo.group(1)
        
        # Check for rust-toolchain.toml channel (if needed, but usually not version-bumped)
        match_toolchain = re.search(r'channel\s*=\s*"(\d+\.\d+\.\d+)"', content)
        if match_toolchain:
            # This is typically rust version, not product version. Handle with care.
            # For this script's purpose, we'll primarily rely on Cargo.toml for product version.
            pass

    raise ValueError(f"Version not found in {file_path}")

def bump_version(current_version: str, bump_type: str) -> str:
    """Bumps the version string based on type (patch, minor, major)."""
    parts = list(map(int, current_version.split('.')))
    if bump_type == 'major':
        parts[0] += 1
        parts[1] = 0
        parts[2] = 0
    elif bump_type == 'minor':
        parts[1] += 1
        parts[2] = 0
    elif bump_type == 'patch':
        parts[2] += 1
    else:
        raise ValueError(f"Invalid bump type: {bump_type}")
    return ".".join(map(str, parts))

def update_file_version(file_path: str, old_version: str, new_version: str):
    """Updates all occurrences of the version string in a file."""
    with open(file_path, 'r') as f:
        content = f.read()

    new_content = content
    
    # Generic version string replacement (e.g., in Cargo.toml version = "X.Y.Z")
    new_content = re.sub(
        r'(version\s*=\s*")' + re.escape(old_version) + r'(")',
        r'\g<1>' + new_version + r'\g<2>',
        new_content
    )
    # Specific UI/doc replacements (e.g., "vX.Y.Z", "MASH INSTALLER vX.Y.Z", "OS REV X.Y.Z")
    # This catches versions in comments, strings, etc.
    new_content = re.sub(
        re.escape(old_version), 
        new_version, 
        new_content
    )
    # For rust-toolchain.toml, if it contains the version directly, though we don't expect it to be auto-bumped this way
    new_content = re.sub(
        r'(channel\s*=\s*")' + re.escape(old_version) + r'(")',
        r'\g<1>' + new_version + r'\g<2>',
        new_content
    )

    if new_content != content:
        print(f"Updating {file_path}: {old_version} -> {new_version}")
        with open(file_path, 'w') as f:
            f.write(new_content)
    else:
        print(f"No change needed in {file_path}")

def run_command(command: str):
    """Runs a shell command and checks for errors."""
    result = subprocess.run(command, shell=True, check=True, capture_output=True, text=True)
    if result.stdout:
        print(result.stdout)
    if result.stderr:
        print(result.stderr)

def main():
    parser = argparse.ArgumentParser(description="Automate version bumping for the MASH Installer workspace.")
    parser.add_argument('bump_type', choices=['patch', 'minor', 'major'],
                        help="Type of version bump to perform.")
    args = parser.parse_args()

    # Get current version from root Cargo.toml
    root_cargo_toml = 'Cargo.toml'
    current_version = get_current_version(root_cargo_toml)
    new_version = bump_version(current_version, args.bump_type)

    print(f"Bumping version from {current_version} to {new_version} ({args.bump_type} bump)")

    workspace_members = get_workspace_members()
    version_files = find_version_files(workspace_members)

    for file_path in version_files:
        update_file_version(file_path, current_version, new_version)
    
    print("\nRunning `cargo update` to update Cargo.lock...")
    run_command("cargo update")
    print("Running `cargo build --workspace` to validate changes and update Cargo.lock if necessary...")
    run_command("cargo build --workspace")
    print("Running `cargo test --workspace` to ensure all tests pass...")
    run_command("cargo test --workspace")

    print(f"\nVersion successfully bumped to {new_version}.")
    print("Remember to update HISTORY.md manually for new entry if this is a release, and commit changes.")

if __name__ == '__main__':
    main()

File: scripts/auto_bump.rs
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

fn get_workspace_members(content: &str) -> Vec<String> {
    let workspace_start = match content.find("[workspace]") {
        Some(p) => p,
        None => return Vec::new(),
    };
    let after = &content[workspace_start..];
    let members_start = match after.find("members") {
        Some(p) => p,
        None => return Vec::new(),
    };
    let after_members = &after[members_start..];
    let open = match after_members.find('[') {
        Some(p) => p,
        None => return Vec::new(),
    };
    let close = match after_members.find(']') {
        Some(p) => p,
        None => return Vec::new(),
    };
    after_members[open + 1..close]
        .split(',')
        .map(|s| s.trim().trim_matches('"').trim().to_string())
        .filter(|s| !s.is_empty() && Path::new(s).is_dir())
        .collect()
}

fn get_current_version(content: &str) -> Option<String> {
    // Find the first `version = "X.Y.Z"` that looks like a semver
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("version") && line.contains('"') {
            if let Some(start) = line.find('"') {
                let rest = &line[start + 1..];
                if let Some(end) = rest.find('"') {
                    let ver = &rest[..end];
                    if ver.split('.').count() == 3
                        && ver.chars().all(|c| c.is_ascii_digit() || c == '.')
                    {
                        return Some(ver.to_string());
                    }
                }
            }
        }
    }
    None
}

fn bump_version(current: &str, bump_type: &str) -> Result<String, String> {
    let parts: Result<Vec<u32>, _> = current
        .split('.')
        .map(|p| p.parse::<u32>())
        .collect();
    let mut parts = parts.map_err(|e| format!("Invalid version '{}': {}", current, e))?;
    if parts.len() != 3 {
        return Err(format!("Expected X.Y.Z, got: {}", current));
    }
    match bump_type {
        "major" => {
            parts[0] += 1;
            parts[1] = 0;
            parts[2] = 0;
        }
        "minor" => {
            parts[1] += 1;
            parts[2] = 0;
        }
        "patch" => {
            parts[2] += 1;
        }
        other => return Err(format!("Unknown bump type: {}. Use patch, minor, or major.", other)),
    }
    Ok(format!("{}.{}.{}", parts[0], parts[1], parts[2]))
}

fn update_file(path: &Path, old: &str, new: &str) -> Result<(), std::io::Error> {
    let content = fs::read_to_string(path)?;
    // Replace `version = "OLD"` entries (Cargo.toml style)
    let updated = content.replace(
        &format!("version = \"{}\"", old),
        &format!("version = \"{}\"", new),
    );
    // Also replace bare version strings in doc/UI files
    let updated = updated.replace(old, new);
    if updated != content {
        println!("  Updating {}: {} -> {}", path.display(), old, new);
        fs::write(path, updated)?;
    } else {
        println!("  No change: {}", path.display());
    }
    Ok(())
}

fn run_command(cmd: &str, args: &[&str]) -> Result<(), String> {
    println!("Running: {} {}", cmd, args.join(" "));
    let status = Command::new(cmd)
        .args(args)
        .status()
        .map_err(|e| format!("Failed to spawn '{}': {}", cmd, e))?;
    if !status.success() {
        return Err(format!(
            "Command `{} {}` failed with exit code {:?}",
            cmd,
            args.join(" "),
            status.code()
        ));
    }
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 || !["patch", "minor", "major"].contains(&args[1].as_str()) {
        eprintln!("Usage: auto_bump <patch|minor|major>");
        std::process::exit(1);
    }
    let bump_type = &args[1];

    let root = env::current_dir()?;
    let root_cargo = root.join("Cargo.toml");
    let root_content = fs::read_to_string(&root_cargo)?;

    let current = get_current_version(&root_content)
        .ok_or("Version not found in root Cargo.toml")?;
    let new_version = bump_version(&current, bump_type)?;

    println!("Bumping version: {} -> {} ({})", current, new_version, bump_type);

    // Collect all Cargo.toml files from workspace members
    let members = get_workspace_members(&root_content);
    let mut files: Vec<PathBuf> = vec![root_cargo];
    for member in &members {
        let member_cargo = root.join(member).join("Cargo.toml");
        if member_cargo.exists() {
            files.push(member_cargo);
        }
    }

    // Add doc and UI files that embed the version string
    for extra in &[
        "docs/MANUAL.md",
        "installer-cli/src/tui/render.rs",
        "installer-cli/src/tui/menus.rs",
        "docs/HISTORY.md",
    ] {
        let p = root.join(extra);
        if p.exists() {
            files.push(p);
        }
    }

    for file in &files {
        update_file(file, &current, &new_version)?;
    }

    println!("\nRunning post-bump validation...");
    run_command("cargo", &["update"])?;
    run_command("cargo", &["build", "--workspace"])?;
    run_command("cargo", &["test", "--workspace"])?;

    println!("\nVersion bumped to {}.", new_version);
    println!("Next: update HISTORY.md with a bardic release entry, then commit and tag.");
    Ok(())
}
